A17-client-side-health-checking.md

```diff
@@ -0,0 +1,317 @@
+receive some number of RPCs that were already in flight from the
+client before the client received the notification that the backend is
+unhealthy.  This race condition lasts approximately the network round
+trip time (it includes the one-way trip time for RPCs that were already
+sent by the client but not yet received by the server when the server
+goes unhealthy, plus the one-way trip time for RPCs sent by the client
+between when the server sends the unhealthy notification and when the
+client receives it).
+
+When the connection is first established, however, the problem could
+affect a larger number of RPCs, because there could be a number of RPCs
+queued up waiting for the channel to become connected, which would all
+be sent at once.  And unlike an already-established connection, the race
+condition is avoidable for newly established connections.
+
+To avoid this, the client will wait for the initial health-checking
+response before the subchannel goes into state `READY`.  However, this
+does mean that when health checking is enabled, we require an additional
+network RTT before the subchannel can be used.  If this becomes a problem
+for users that cannot be solved by simply disabling health-checking,
+we can consider adding an option in the future to treat the subchannel
+as healthy until the initial health-checking response is received.
+
+#### Call Credentials
+
+If there are any call credentials associated with the channel, the client
+will send those credentials with the `Watch()` call.  However, we will not
+provide a way for the client to explicitly add call credentials for the
+`Watch()` call.
+
+### Service Config Changes
+
+We will add the following new message to the service config:
+
+```
+message HealthCheckConfig {
+  // Service name to use in the health-checking request.
+  google.protobuf.StringValue service_name = 1;
+}
+```
+
+We will then add the following top-level field to the ServiceConfig proto:
+
+```
+HealthCheckConfig health_check_config = 4;
+```
+
+Note that we currently need only this one parameter, but we are putting
+it in its own message so that we have the option of adding more parameters
+later if needed.
+
+### LB Policies Can Disable Health Checking When Needed
+
+There are some cases where an LB policy may want to disable client-side
+health-checking.  To allow this, LB policies will be able to set the
+channel argument mentioned above to inhibit health checking.
+
+This section details how each of our existing LB policies will interact
+with health checking.
+
+#### `pick_first`
+
+We do not plan to support health checking with `pick_first`, because it
+is not clear what the behavior of `pick_first` would be for unhealthy
+channels.  The naive approach of treating an unhealthy channel as having
+failed and disconnecting would both result in expensive reconnection
+attempts and might not actually cause us to select a different backend
+after re-resolving.
```

Comment:
Please add more details here. Intuitively, it seems like pick-first policy should skip the address if it is not healthy and pick the next healthy one. The issue of expensive reconnection attempts (subject to exponential backoff) also exists for RR policy, right?

Reply:
No, this issue does not exist for RR, because RR never disconnects from a backend -- it will always be attempting to connect to all backend addresses in its list.  So if a backend connection fails in RR, RR just keeps trying.  From the subchannel's point of view, that does not cause any connection churn, regardless of whether the subchannel is actually disconnected or is connected but reporting unhealthy.

In contrast, in PF, if a backend connection fails, PF gives up on it and goes back to its list of addresses to try to find one that will work.  That's why this is harder for PF -- it would actually cause us to try to switch to a different address, which would cause a lot of connection churn.